Extensions/ParticleSystem/particleemitterobject.ts

```diff
@@ -680,75 +653,76 @@ namespace gdjs {
       }
     }
 
-    getParticleSize2() {
+    getParticleSize2(): float {
       return this.size2;
     }
 
-    setParticleSize2(size): void {
+    setParticleSize2(size: float): void {
       if (this.size2 !== size) {
         this._sizeDirty = true;
         this.size2 = size;
       }
     }
 
-    getParticleAlpha1() {
+    getParticleAlpha1(): number {
       return this.alpha1;
     }
 
-    setParticleAlpha1(alpha): void {
+    setParticleAlpha1(alpha: number): void {
       if (this.alpha1 !== alpha) {
         this._alphaDirty = true;
         this.alpha1 = alpha;
       }
     }
 
-    getParticleAlpha2() {
+    getParticleAlpha2(): number {
       return this.alpha2;
     }
 
-    setParticleAlpha2(alpha): void {
+    setParticleAlpha2(alpha: number): void {
       if (this.alpha2 !== alpha) {
         this._alphaDirty = true;
         this.alpha2 = alpha;
       }
     }
 
-    noMoreParticles() {
-      this._renderer.stop();
+    noMoreParticles(): boolean {
+      return !this._renderer.emitter.emit;
     }
 
-    recreateParticleSystem() {
+    recreateParticleSystem(): void {
       this._renderer.recreate();
     }
 
-    getFlow() {
+    getFlow(): number {
       return this.flow;
     }
 
-    setFlow(flow): void {
+    setFlow(flow: number): void {
       if (this.flow !== flow) {
         this.flow = flow;
         this._flowDirty = true;
       }
     }
 
-    getParticleCount() {
+    getParticleCount(): number {
       return this._renderer.getParticleCount();
     }
 
-    getTank() {
+    getTank(): number {
       return this.tank;
     }
 
-    setTank(tank): void {
+    setTank(tank: number): void {
       this.tank = tank;
+      this._tankDirty = true;
```

Comment:
nit: let's do a `if (this.tank !== tank) {` to avoid marking the tank as dirty if that's not the case. More practical for users not to have to check this by themselves, especially because this reset the emission. Otherwise, an event like:

Conditions : none
Actions: Set the tank of MyParticleEmitter to 1000

would result in a "perpetually resetting emission" (so no particles emitted), which would be considered as a bug/confusing behavior for users.

Reply:
I tried but I saw an issue with that: one could set the tank to the same capacity as previously to have it emit the same quantity. The issue is that `this.tank` is actually the initial tank capacity, not the current one.

The event that you took as an example would actually, on each step, recompute the emitter lifetime and would emit until the end of times. It wouldn't be that unexpected to me.

Reply:
> The issue is that this.tank is actually the initial tank capacity, not the current one.

This is very true!

> The event that you took as an example would actually, on each step, recompute the emitter lifetime and would emit until the end of times. It wouldn't be that unexpected to me.

Could be indeed not unexpected... could we make a simple test game to see the behavior? I was worried resetting the emission would also delete the existing particles, which would result in the player seeing nothing on screen (because every frame the tank is changed and the emission is reset - but I'm not sure if this is what happens).

Reply:
![particleEmitterInfiniteTest](https://user-images.githubusercontent.com/32449369/134299663-58013852-6150-4361-8b5a-0fd24dd82418.gif)

Here it is!

For the record, we could have an expression that returns the number of emitted/remaining particles but it cannot be reliable for the following reason:

- Pixi does not have the concept of tank/capacity that is present in GDevelop, we do the conversion to frequency and lifetime that pixi actually uses.
- On update, pixi will check if it has to emit particles comparing elapsed time to the period of emission.
- If it finds out particles should be emitted, it will first check that there is not too much particles already displayed (via the max parameter).
- If indeed the maximum has been reached, it does not emit more particle and there is no mechanism to say "We should have emitted X particles but we couldn't, so we will keep a record for that and emit them later, extending the emitter lifetime."

I don't know if this should be explained in the editor.

Reply:
Thanks for checking and for the GIF. I think it's fine then as you did, and that as long as the emission is done and visible, it's fine to "reset" it when the tank is modified.

If that generates confusion in the future, we can always hide the action to modify the tank and implement new ones that deals with the emitter lifetime/frequency directly.
But for now I think it's pretty good like this then!